tex/generic/pgf/frontendlayer/tikz/tikz.code.tex

```diff
@@ -2123,125 +2123,43 @@
-        \let\pgfutil@next=\tikz@g@char%
-      \else%
-        \ifx\pgf@let@token s%
-           \let\pgfutil@next=\tikz@schar%
-        \else%
-          \ifx\pgf@let@token |%
-             \let\pgfutil@next=\tikz@vh@lineto%
-          \else%
-            \ifx\pgf@let@token p%
-              \let\pgfutil@next=\tikz@pchar%
-              \pgfsetmovetofirstplotpoint%
-            \else%
-              \ifx\pgf@let@token t%
-                \let\pgfutil@next=\tikz@to%
-              \else%
-                \ifx\pgf@let@token\pgfextra%
-                  \let\pgfutil@next=\tikz@extra%
-                \else%
-                  \ifx\pgf@let@token\foreach%
-                    \let\pgfutil@next=\tikz@foreach%
-                  \else%
-                    \ifx\pgf@let@token f%
-                      \let\pgfutil@next=\tikz@fchar%
-                    \else%
-                      \ifx\pgf@let@token\pgf@stop%
-                        \let\pgfutil@next=\relax%
-                      \else%
-                        \ifx\pgf@let@token\par%
-                          \let\pgfutil@next=\tikz@scan@next@command%
-                        \else%
-                          \ifx\pgf@let@token d%
-                            \let\pgfutil@next=\tikz@decoration%
-                          \else%
-                            \ifx\pgf@let@token l%
-                              \let\pgfutil@next=\tikz@l@char%
-                            \else%
-                              \ifx\pgf@let@token:%
-                                \let\pgfutil@next=\tikz@colon@char%
-                              \else%
-                                \let\pgfutil@next=\tikz@expand%
-                              \fi%
-                            \fi%
-                          \fi%
-                        \fi%
-                      \fi%
-                    \fi%
-                  \fi%
-                \fi%
-              \fi%
-            \fi%
-          \fi%
-        \fi%
-      \fi%
-    \fi%
-  \fi%
-  \ifx\pgfutil@next\tikz@expand\else\tikz@expandcount=100\relax\fi%
-  \pgfutil@next%
+  \pgfutil@switch\pgfutil@ifx\pgf@let@token{%
+    {(}{\tikz@movetoabs}%)
+    {+}{\tikz@movetorel}%
+    {-}{\tikz@lineto}%
+    {.}{\tikz@dot}%
+    {r}{\tikz@rect}%
+    {n}{\tikz@fig}%
+    {[}{\tikz@parse@options}%]
+    {c}{\tikz@cchar}%
+    {\bgroup}{\tikz@beginscope}%
+    {\egroup}{\tikz@endscope}%
+    {;}{\tikz@finish}%
+    {a}{\tikz@a@char}%
+    {e}{\tikz@e@char}%
+    {g}{\tikz@g@char}%
+    {s}{\tikz@schar}%
+    {|}{\tikz@vh@lineto}%
+    {p}{\pgfsetmovetofirstplotpoint\tikz@pchar}%
+    {t}{\tikz@to}%
+    {\pgfextra}{\tikz@extra}%
+    {\foreach}{\tikz@foreach}%
+    {f}{\tikz@fchar}%
+    {\pgf@stop}{\relax}%
+    {\par}{\tikz@scan@next@command}%
+    {d}{\tikz@decoration}%
+    {l}{\tikz@l@char}%
+    {:}{\tikz@colon@char}%
+    {\relax}{\relax\tikz@scan@next@command}%
+    {\tikz@frozen@relax@token}{\tikz@frozen@relax@token\tikz@scan@next@command}%
```

Comment:
Do you think we should handle frozen relax in this way?  This only happens when TeX failed to read a number which means that the statement that wanted to use the number becomes nonsensical.  I think this will be even more confusing for users than `Giving up on this path. Did you forget a semicolon?`

Reply:
Reasonable. Having `\relax` handled is enough for inputs like `\draw (0,0) \ifnum0=0\relax\fi -- (1,1);` with "standard" texniques applied.  I will remove the frozen `\relax` part. TeX is just too dynamic/free to not define DSLs one after another but too weak to assert user input and catch errors.